# RNAseq Data - GSE189984

## General info
This dataset (GSE189984) was generated in 2022 and aimed to identify genes regulated by Vit D in the human peripheral blood mononuclear (PBMC) cell line. Expression profiling was performed through high throughput RNA sequencing after treatment with Vit D3 for 4 / 8 / 24 / 48 h. This study was performed on 24 samples (in triplicate form).

- Read length: 75bp
- Single/paired end sequencing: single end sequencing
- Platform used: Illumina NextSeq500

Due to time constraints and HPC downtime, already normalized count tables were used instead of aligning the reads ourselves.

## Analysis

### Loading Libraries

```{r echo=FALSE, results='hide', error=FALSE, warning=FALSE, message=FALSE}
# install packages if not yet installed.
pkgs <- c("dplyr", "edgeR", "biomaRt", "ggplot2")
notInstalled <- pkgs[!pkgs %in% installed.packages()[,1]]
if(length(notInstalled) > 0){
  BiocManager::install(notInstalled)
}

library(dplyr)
library(edgeR)
library(biomaRt)
library(ggplot2)
```

### Normalized count table

The already normalized RNAseq count matrix is loaded. The column names of this matrix also contains information about the sample and corresponding treatment.

```{r}
# count data
counts <- read.csv("./data/RNAseq189984/GSE189984_counts.csv", header = TRUE, row.names = 1)
counts
```

### Preprocessing

#### Duplicates

Check for duplicate gene counts?

TODO: check video

```{r}
dups = duplicated(rownames(counts))
sum(dups)
```

TODO: sum technical replicates?

#### Annotation

The transcript annotation data is extracted. This contains gene symbol, ID, description, and other less relevant information (e.g. chromosome).

```{r}
# annotation data
annot <- read.csv("./data/RNAseq189984/GSE189984_annot.csv", header = TRUE)
head(annot)
```

A case-control study was performed with 4 different treatment groups (4, 8, 24 and 48 hour exposure) in the PBMC cell line. The treatment time will be added as a factor during statistical analysis.

```{r}
# extract sample information necessary for generating the design matrix
annotations <- strsplit(colnames(counts), split = "_")

annotation <- data.frame(
  "treatment" = factor(sapply(annotations, function(x) x[1])),
  "time" = factor(sapply(annotations, function(x) x[2])),
  "replicate" = factor(sapply(annotations, function(x) x[3])),
  row.names = counts_cols)
#colnames(annotation) <- c("treatment", "time", "replicate")
annotation
```

Check if the colnames match up in order with the annotation data.

```{r}
rownames(annotation) == colnames(counts)
```

## EdgeR Analysis
TODO: in the practicum we did more steps to accomodate the counts into EdgeR.


```{r}
dge <- DGEList(counts, group = annotation$treatment)
head(dge$sample)
```

### Design Matrix

```{r}
design <- model.matrix(~ treatment * time, annotation)
rownames(design) <- colnames(dge)
head(design, 2)
```

### Filtering

```{r}
keep <- filterByExpr(dge, design)
print(table(keep))
dge <- dge[keep, , keep.lib.sizes = FALSE]
```

### Normalisation

```{r}
dge <- calcNormFactors(dge)
head(dge$samples)
```

### Data exploration

MDS plot

```{r}
limma::plotMDS(
  dge, 
  labels = paste(annotation$treatment, annotation$time, sep=" - "), 
  col = as.double(annotation$treatment))
```

Observations:
- maybe no time effect for Ethanol treated samples? but there does seem to be clustering for 4/8 and 24/48
- time effect for VitD treated samples (expected), clear clustering also shown


### Estimate dispersion

```{r}
dge <- estimateDisp(dge, design)  
plotBCV(dge)
```

Looks OK? -> TODO: why, not a real trend

### Fit quasi-negative binomial model

```{r}
fit <- glmQLFit(dge,design)
```

### Define contrasts

TODO: do the contrasts properly

```{r}
L <- msqrob2::makeContrast(
    c("treatmentX125D = 0",
      "time08h = 0",
      "time24h = 0",
      "time48h = 0",
      "treatmentX125D:time08h = 0",
      "treatmentX125D:time24h = 0",
      "treatmentX125D:time48h = 0",
      "treatmentX125D + treatmentX125D:time08h = 0",
      "treatmentX125D + treatmentX125D:time24h = 0",
      "treatmentX125D + treatmentX125D:time48h = 0"),
    parameterNames = colnames(design))
head(L, 2)
```

Perform Contrast Tests

```{r}
testsF <- apply(L, 2, function(fit, contrast) 
    glmQLFTest(fit, contrast = contrast), 
    fit = fit)

topTablesF<- lapply(testsF, topTags, n = nrow(dge))

sig_counts <- sapply(topTablesF, function(x) sum(x$table$FDR< 0.05))
sig_counts
```

Expected, there shouldn't really be a time effect? (as Ethanol does nothing?)

### Volcano Plots

```{r}
for (i in 1:ncol(L)){
  volcano <- ggplot(topTablesF[[i]]$table, 
                    aes(x = logFC, y = -log10(PValue), color = FDR < 0.05)) + geom_point() + scale_color_manual(values = c("black","red")) + ggtitle(paste("contrast", names(topTablesF)[i]))
  print(volcano)
}
```

### Histograms of P-values

```{r}
histsP <- lapply(topTablesF, function(x) 
    x$table %>% 
        ggplot(aes(x = PValue)) + 
        geom_histogram(breaks = seq(0,1,.1) , col = 1)
    )
    
for (i in 1:ncol(L)) 
    histsP[[i]] <- histsP[[i]] +
    ggtitle(paste("contrast", names(topTablesF)[i]))
histsP
```

TODO: some of these plots might seem a bit weird.

### Checks

Convert gene symbol into ensembl id.

```{r}
symbol2ensembl = function(x) {
  return(data[data$external_gene_name == x, "ensembl_gene_id"][1])
}
```

```{r}
genes4h <- topTablesF$treatmentX125D$table
genes8h <- topTablesF$`treatmentX125D + treatmentX125D:time4h`$table
genes24h <- topTablesF$`treatmentX125D + treatmentX125D:time24h`$table
genes48h <- topTablesF$`treatmentX125D + treatmentX125D:time48h`$table
```

TODO: actual interpretation

#### Percentages upregulated vs downregulated genes

TODO: why the mismatch?

```{r}
sig4 = genes4h %>% filter(FDR < 0.05)
c4up = sig4 %>% filter(logFC > 0) %>% count
c4 = sig4 %>% count
print(c4up / c4)

sig8 = genes8h %>% filter(FDR < 0.05)
c8up = sig8 %>% filter(logFC > 0) %>% count - c4up
c8 = sig8 %>% count - c4
print(c4)
print(c4up / c4)

sig24 = genes24h %>% filter(FDR < 0.05)
c24up = sig24 %>% filter(logFC > 0) %>% count - c8up - c4up
c24 = sig24 %>% count - c8 - c4
print(c24)
print(c24up / c24)

sig48 = genes48h %>% filter(FDR < 0.05)
c48up = sig48 %>% filter(logFC > 0) %>% count - c24up - c8up - c4up
c48 = sig24 %>% count - c24 - c8 - c4
print(c48)
print(c48up / c48)
```

#### Number of genes that only become significant after 24h
In the paper they mention that only after 24h of treatment, 995 genes become significantly expressed.

We count:

```{r}
at_24h = sig_counts["treatmentVitD + time24h:treatmentVitD + time4h:treatmentVitD"]
at_4h = sig_counts["treatmentVitD + time4h:treatmentVitD"]

at_24h - at_4h
```

#### NOD2

According to this paper (*Epigenome-wide effects of vitamin D and their impact on the transcriptome of human monocytes involve CTCF*),
expression of this gene ENSG00000167207 (NOD2), is only affected after 24h .

Evaluate this.

```{r}
gene = "ENSG00000167207"

genes2.5h[gene,]$FDR
genes4h[gene,]$FDR
genes24h[gene, ]$FDR
```
Indeed, only significant after 24h of treatment.

#### HTT
HTT: ENSG00000197386

Should become significant after 4h of treatment.

```{r}
gene = "ENSG00000197386"



genes2.5h[gene,]$FDR
genes4h[gene,]$FDR
genes24h[gene, ]$FDR
```
Indeed, only significant after 4h of treatment.

 
#### THBD

Significant gene from the GSE5145 microarray dataset.

- THBD: ENSG00000178726
- PEAK1: 

```{r}


gene <- symbol2ensembl("THBD")
gene

genes2.5h[gene,]$FDR
genes4h[gene,]$FDR
genes24h[gene, ]$FDR


gene <- symbol2ensembl("PEAK1")
gene

genes2.5h[gene,]$FDR
genes4h[gene,]$FDR
genes24h[gene, ]$FDR


gene <- symbol2ensembl("SULT1C2")
gene

genes2.5h[gene,]$FDR
genes4h[gene,]$FDR
genes24h[gene, ]$FDR

```

### Comparisons
#### With microarray data

```{r}
# go over all the genes from the microarray dataset.
# TODO: get the significant genes

counts <- 0

for (symbol in df$Gene.symbol) {
  id <- symbol2ensembl(symbol)
  
  if (!(id %in% rownames(genes24h))) {
    next
  }
  
  if(genes24h[id, ]$FDR < 0.05) {
    counts <- counts + 1
  }
}

counts
```

#### With other RNAseq dataset

```{r}
# go over all the genes from the microarray dataset.
# TODO: get the significant genes

counts <- 0

for (gene in rownames(result)) {

  if (!(gene %in% rownames(genes24h))) {
    next
  }
  
  if(genes24h[gene, ]$FDR < 0.05) {
    counts <- counts + 1
    print(gene)
  }
}

counts
```