---
title: "Appendix D: RNAseq Analysis (GSE189984)"
output:
  pdf_document:
    df_print: kable
---

```{r setup, include=FALSE} 
# disable warnings & mes
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
knitr::opts_chunk$set(dpi = 120)
```


# RNAseq Data - GSE189984

## General info

This dataset was generated in 2022 and aimed to identify genes regulated by Vit D in the human peripheral blood mononuclear (PBMC) cell line. Expression profiling was performed through high throughput RNA sequencing after treatment with Vit D3 for 4 / 8 / 24 / 48 h. This study was performed on 24 samples (in triplicate form).

- Read length: 75bp
- Single/paired end sequencing: single end sequencing
- Platform used: Illumina NextSeq 500

> Due to time constraints and HPC downtime, normalized count tables already mapped to a genome were used instead of aligning the reads ourselves.

The orginal paper by Hanel & Carlberg (2022) includes some general information on quality control and read mapping: Reads were aligned to the reference genome (version GRCh38) with Ensembl annotation (version 103) by using default settings of the nf-core/rnaseq STAR-Salmon pipeline (version 3.0). Ensembl gene identifiers were annotated with gene symbol, description, genomic location and biotype by accessing the Ensembl database (version 104) via the R package BiomaRt (version 2.46.0). Ensembl gene identifiers missing HGNC gene symbol annotation, Entrez ID, genomic location information or being mitochondrially encoded were removed from the data sets. When a gene name appeared more than once, the entry with the highest average gene counts was kept. Read counts were normalized for differences in library size to counts per million (CPM). Genes with very low expression were filtered out by applying the function FilterByExpr(), in order to mitigate the multiple testing problem and to not interfere with the statistical approximations of the EdgeRpipeline. This requirement was fulfilled by 12,305 genes. After filtering, library sizes were recomputed and trimmed mean of M-value normalization was applied.

## Preparation

### Loading Libraries

```{r echo=FALSE, results='hide', error=FALSE, warning=FALSE, message=FALSE}
# install packages if not yet installed.
pkgs <- c("dplyr", "edgeR", "biomaRt", "ggplot2")
notInstalled <- pkgs[!pkgs %in% installed.packages()[,1]]
if(length(notInstalled) > 0){
  BiocManager::install(notInstalled)
}

library(dplyr)
library(edgeR)
library(biomaRt)
library(ggplot2)
```

### Loading Count Table

The already normalized RNAseq count matrix is loaded. The column names of this matrix contains information about the sample and corresponding treatment.

```{r}
# count data
counts <- read.csv("./data/RNAseq189984/GSE189984_counts.csv", header = TRUE, row.names = 1)
print(paste("Count table dimensions:", dim(counts)[1], 'x', dim(counts)[2], sep = ' '))

# number of columns reduced for printing
head(counts[,1:5])
```

The count table does not contain any duplicate genes.

```{r}
dups = duplicated(rownames(counts))
sum(dups)
```

### Load Annotation Matrix

The transcript annotation data is extracted. This contains gene symbol, ID, description, and other less relevant information (e.g. chromosome).

```{r}
# annotation data
annot <- read.csv("./data/RNAseq189984/GSE189984_annot.csv", header = TRUE)
head(annot[,c("hgnc_symbol", "ensembl_gene_id", "entrezgene_id", "description")])
```

A case-control study was performed with 4 different treatment groups (4, 8, 24 and 48 hour exposure) in the PBMC cell line. The treatment time will be added as a factor during statistical analysis.

```{r}
# extract sample information necessary for generating the design matrix
annotations <- strsplit(colnames(counts), split = "_")

annotation <- data.frame(
  "treatment" = factor(sapply(annotations, function(x) x[1])),
  "time" = factor(sapply(annotations, function(x) x[2])),
  "replicate" = factor(sapply(annotations, function(x) x[3])),
  row.names = colnames(counts))

head(annotation)
```

## EdgeR Statistical Analysis

An EdgeR object is generated with treatment as a grouping factor.

```{r}
dge <- DGEList(counts, group = annotation$treatment)
head(dge$sample)
```

Some differences in library size are observed, this will be accounted for during normalization.

### Design Matrix

A design matrix is generated, taking into account treatment and treatment time.

```{r}
design <- model.matrix(~ treatment * time, annotation)
rownames(design) <- colnames(dge)
data.frame(head(design))
```

### Normalisation

Normalization is performed with the default `calcNormFactors()` function of the EdgeR package.

```{r}
dge <- calcNormFactors(dge)
dge$samples
```

### Filtering

Filtering is performed with the default `filterByExpr()` function of the EdgeR package.

```{r}
keep <- filterByExpr(dge, design)
table(keep)
dge <- dge[keep, , keep.lib.sizes = FALSE]
```

### Data Exploration

An MDS plot, showing sample grouping is generated. A clear distinction between treatments (Vit D or EtOH) and treatment times (4, 8, 24, and 48 h) can be observed. Notably, the untreated samples also seem to cluster based on treatment time.

```{r}
limma::plotMDS(
  dge, 
  labels = paste(annotation$treatment, annotation$time, sep=" - "), 
  col = as.double(annotation$treatment))
```


### Estimate Dispersion

The dispersion is estimated and a genewise biological coefficient of variation (BCV) plot is generated. Only a very slight trend is observed.

```{r}
dge <- estimateDisp(dge, design)  
plotBCV(dge)
```

### Model Fit

A quasi-negative binomial model is fitted, and a QQ-plot is generated to assess the goodness of fit. Some disturbances in spread can be observed, especially at the higher quantiles.

```{r}
fit <- glmQLFit(dge, design)
gof(fit, plot = TRUE)
```

### Contrast Testing

A contrast matrix is generated to test significance for any of the selected variables.

```{r}
L <- msqrob2::makeContrast(
    c("treatmentX125D = 0",
      "time08h = 0",
      "time24h = 0",
      "time48h = 0",
      "treatmentX125D:time08h = 0",
      "treatmentX125D:time24h = 0",
      "treatmentX125D:time48h = 0",
      "treatmentX125D + treatmentX125D:time08h = 0",
      "treatmentX125D + treatmentX125D:time24h = 0",
      "treatmentX125D + treatmentX125D:time48h = 0"),
    parameterNames = colnames(design))
head(L, 3)
```

Contrast tests are performed and the number of significant values (after multiple testing correction) is determined. Some significant values can be obtained for every contrast that was tested for.

```{r}
testsF <- apply(L, 2, function(fit, contrast) 
    glmQLFTest(fit, contrast = contrast), 
    fit = fit)

topTablesF <- lapply(testsF, topTags, n = nrow(dge))

sig_counts <- sapply(topTablesF, function(x) sum(x$table$FDR < 0.05))
sig_counts
```


```{r}
genes4h <- topTablesF[["treatmentX125D"]][["table"]]
genes8h <- topTablesF[["treatmentX125D + treatmentX125D:time08h"]][["table"]]
genes24h <- topTablesF[["treatmentX125D + treatmentX125D:time24h"]][["table"]]
genes48h <- topTablesF[["treatmentX125D + treatmentX125D:time48h"]][["table"]]
```

### Assesment of Significance

Volcano plots are generated to assess significance and general direction of differential expression. Vit D treatment seems to stimulate gene expression, and the time effect seems to play quite and important role (even in the absence of Vit D treatment).

```{r}
for (i in 1:ncol(L)){
  volcano <- ggplot(topTablesF[[i]]$table, 
                    aes(x = logFC, y = -log10(PValue), color = FDR < 0.05)) +
      geom_point() +
      scale_color_manual(values = c("black","red")) +
      ggtitle(paste("contrast", names(topTablesF)[i]))
  print(volcano)
}
```

P-value histograms are generated to assess the distribution of significant findings for each contrast tested. Once again, it is shown that time effect plays a significant role in the differentiation of expression.

```{r}
histsP <- lapply(topTablesF, function(x) 
    x$table %>% 
        ggplot(aes(x = PValue)) + 
        geom_histogram(breaks = seq(0, 1, .1) , col = 1)
    )
    
for (i in 1:ncol(L)) 
    histsP[[i]] <- histsP[[i]] +
    ggtitle(paste("Histogram of P-values\nfor contrast", names(topTablesF)[i]))
histsP
```

### MA plots

Significant DE between non-treated and treated cells at 24h time point.

```{r}
with(genes24h,plot(logCPM,logFC,pch=16,cex=0.2))
# MAplot: all data points
with(genes24h,points(logCPM[FDR<0.05],logFC[FDR<0.05],pch=16,col="red",cex=0.6))
# MA-plot: significant loci
abline(0,0)
```

### Add Gene Symbols:

```{r}
# annotation data
GRCh38 <-  useEnsembl(biomart="genes", host="https://www.ensembl.org",
                      dataset="hsapiens_gene_ensembl")

# attributes
atr <- listAttributes(GRCh38)
data <- getBM(attributes = c('ensembl_gene_id', 'ensembl_transcript_id',
                             'external_gene_name'), mart = GRCh38)

symb_f <- data[!duplicated(data$ensembl_gene_id),]

add_gene_symbols <- function (genes) {
    symb_f <- symb_f[symb_f$ensembl_gene_id %in% rownames(genes),]
    symb_f <- symb_f[order(symb_f$ensembl_gene_id),]
    genes_sort <- genes[rownames(genes4h) %in% symb_f$ensembl_gene_id,] # not all ensembl ids are in the reference
    genes_sort <- genes_sort[order(rownames(genes_sort)),]
    genes_sort$gene_symbol <- symb_f$external_gene_name

    return(genes_sort)
}

genes4h_sort <- add_gene_symbols(genes4h)
genes8h_sort <- add_gene_symbols(genes8h)
genes24h_sort <- add_gene_symbols(genes24h)
genes48h_sort <- add_gene_symbols(genes48h)
```

### Top hits at 24h

```{r}
genes48h_top <- genes48h_sort[order(genes48h_sort$FDR),]
top <- head(genes48h_top[,c("gene_symbol", "logFC", "FDR")])
top
```

### Save Results

```{r}
write.csv(genes4h_sort, "results/GSE189984/genes4h.csv")
write.csv(genes8h_sort, "results/GSE189984/genes8h.csv")
write.csv(genes24h_sort, "results/GSE189984/genes24h.csv")
write.csv(genes48h_sort, "results/GSE189984/genes48h.csv")

```


### Summarizing & Annotating Results

The gene symbol has to be converted into Ensembl gene id.

```{r}
symbol2ensembl = function(x) {
  return(data[data$external_gene_name == x, "ensembl_gene_id"][1])
}
```

Significant genes for each treatment time are extracted and filtered. The fraction of up/down-regulated genes is determined for each of these.

```{r}
sig4 = genes4h %>% filter(FDR < 0.05)
c4up = sig4 %>% filter(logFC > 0) %>% count
c4 = sig4 %>% count
print(paste("Number of significant genes when treated after 4h:", c4[,1]))
print(paste("Fraction of up-regulated genes when treated after 4h:", (c4up / c4)[,1]))

sig8 = genes8h %>% filter(FDR < 0.05)
c8up = sig8 %>% filter(logFC > 0) %>% count
c8 = sig8 %>% count
print(paste("Number of significant genes when treated after 8h:", c8[,1]))
print(paste("Fraction of up-regulated genes when treated after 8h:", (c8up / c8)[,1]))

sig24 = genes24h %>% filter(FDR < 0.05)
c24up = sig24 %>% filter(logFC > 0) %>% count
c24 = sig24 %>% count
print(paste("Number of significant genes when treated after 24h:", c24[,1]))
print(paste("Fraction of up-regulated genes when treated after 24h:", (c24up / c24)[,1]))

sig48 = genes48h %>% filter(FDR < 0.05)
c48up = sig48 %>% filter(logFC > 0) %>% count
c48 = sig48 %>% count
print(paste("Number of significant genes when treated after 48h:", c48[,1]))
print(paste("Fraction of up-regulated genes when treated after 48h:", (c48up / c48)[,1]))
```


