# RNAseq Data - GSE69284

## General info
This dataset (GSE69284) was generated in 2016 and aimed to identify genes regulated by Vit D in the human monocytic THP-1 cell line. Expression profiling was performed through high throughput RNA sequencing after treatment with Vit D3 for 2.5 / 4 / 24 h. This study was performed on 18 samples (in triplicate form), 4 of which were used for our data analysis.

- Read length: 51bp
- Single/paired end sequencing: single end sequencing
- Platform used: Illumina NextSeq500

## Quality control
A MultiQC quality control was performed and mapping statistics are provided through the kallisto log output.

TODO

```{r}

```


```{r, echo = FALSE, results='asis'}
htmltools::includeHTML("./stats/RNAseq/GSE69284/2_multiqc/multiqc_report.html")
```

## Analysis

### Loading Libraries

```{r echo = T, results = 'hide', error=FALSE, warning=FALSE, message=FALSE}
# install packages if not yet installed.
pkgs <- c("txmport", "dplyr", "edgeR", "biomaRt")
notInstalled <- pkgs[!pkgs %in% installed.packages()[,1]]
if(length(notInstalled) > 0){
  BiocManager::install(notInstalled)
}

library(tximport)
library(dplyr)
library(edgeR)
library(biomaRt)
```

### Pseudo count table
Kallisto transcript mapping was performed to generate a pseudo-count table.

The transcript annotation data is extracted.

```{r}
# annotation data
GRCh38 <-  useEnsembl(biomart="genes", host="https://www.ensembl.org", dataset="hsapiens_gene_ensembl")

# attributes
atr <- listAttributes(GRCh38)
atr
data <- getBM(attributes = c('ensembl_gene_id', 'ensembl_transcript_id', 'external_gene_name'), mart = GRCh38)

tx2geneGtf <- dplyr::select(data, ensembl_transcript_id, ensembl_gene_id)

# maps transcript -> gene
tx2geneGtf <- rename(tx2geneGtf, TXNAME = ensembl_transcript_id)
tx2geneGtf <- rename(tx2geneGtf, GENEID = ensembl_gene_id)
head(tx2geneGtf)
```

The Kallisto pseudoalignment data is loaded.

Search for the Kallisto output files.

```{r}

files <- list.files("./stats/RNAseq/GSE69284", recursive=TRUE)

abundance_files <- grep("abundance_bis.tsv", files, value=TRUE)
samples <- gsub("(.*)_quant.*", "\\1", abundance_files, perl=TRUE)
abundance_files <- paste(rep("./stats/RNAseq/GSE69284/"), abundance_files, sep="")

names(abundance_files) <- samples
abundance_files
```

Load RNAseq data

```{r}
txi <- tximport(abundance_files, type = "kallisto", tx2gene = tx2geneGtf)

head(txi$counts)
dim(txi$counts)
```

### Preprocessing

#### Duplicates

Check for duplicate gene counts?
TODO: check video

```{r}
dups = duplicated(rownames(txi$counts))
sum(dups)
```

TODO: sum technical replicates?

#### Annotation

General sample annotation file was provided.  
A case-control study was performed with 3 different treatment groups (2.5, 4, and 24 hour exposure) in the THP-1 cell line. The treatment time will be added as a factor during statistical analysis.
```{r}
RNAseq_annot <- read.csv("./data/annotation/GSE69284/SraRunTable.txt")
colnames(RNAseq_annot)

annotation <- RNAseq_annot[c("TREATMENT", "Treatment_Time")]
colnames(annotation) <- c("treatment", "time")
rownames(annotation) <- RNAseq_annot$Run

# convert into factors
annotation$treatment <- factor(annotation$treatment, levels=c("Ethanol", "100 nM 1\\,25(OH)2D3"), labels=c("Ethanol", "VitD"))
annotation$time <- factor(annotation$time, levels=c("2.5 h", "4 h", "24 h"), labels=c("2.5h", "4h", "24h"))

annotation
```

Check if the colnames match up in order with the annotation data.

```{r}
rownames(annotation) == colnames(txi$counts)
```

## EdgeR Analysis
TODO: in the practicum we did more steps to accomodate the counts into EdgeR.


```{r}
dge <- DGEList(counts=txi$counts)
dge$sample
```

### Design Matrix

```{r}
design <- model.matrix(~time * treatment, annotation)
rownames(design) = colnames(dge)
design
```

### Filtering

```{r}
keep <- filterByExpr(dge,design)
print(table(keep))
dge <- dge[keep, , keep.lib.sizes=FALSE]
```
### Normalisation

```{r}
dge <- calcNormFactors(dge)
dge$samples
```

### Data exploration

MDS plot

```{r}
limma::plotMDS(dge,labels=paste(annotation$treatment,annotation$time,sep=" - "),col=as.double(annotation$treatment))
```

Observations:
- maybe no time effect for Ethanol treated samples? (expected)
- time effect for VitD treated samples (expected)


### Estimate dispersion

```{r}
dge <- estimateDisp(dge, design)  
plotBCV(dge)
```

Looks OK? -> TODO: why

### Fit quasi-negative binomial model

```{r}
fit <- glmQLFit(dge,design)
```

### Define contrasts

TODO: do the contrasts properly

```{r}
L <- msqrob2::makeContrast(
    c("treatmentVitD = 0",
      "time24h:treatmentVitD = 0",
      "time4h:treatmentVitD = 0",
      "treatmentVitD + time4h:treatmentVitD = 0",
      "treatmentVitD + time24h:treatmentVitD + time4h:treatmentVitD = 0",
      "time4h = 0",
      "time24h = 0"),
    parameterNames = colnames(design))
L
```
Perform Contrast Tests

```{r}
testsF <- apply(L, 2, function(fit,contrast) 
    glmQLFTest(fit,contrast=contrast), 
    fit = fit)

topTablesF<- lapply(testsF, topTags, n=nrow(dge))

sig_counts = sapply(topTablesF, function(x) sum(x$table$FDR< 0.05))
sig_counts
```

Expected, there shouldn't really be a time effect? (as Ethanol does nothing?)

### Volcano Plots

```{r}
for (i in 1:ncol(L))
{
   volcano<- ggplot(topTablesF[[i]]$table,aes(x=logFC,y=-log10(PValue),color=FDR < 0.05)) + geom_point() + scale_color_manual(values=c("black","red")) + ggtitle(paste("contrast",names(topTablesF)[i]))
print(volcano)
}
```

### Histograms of P-values
```{r}
histsP <- lapply(topTablesF, function(x) 
    x$table %>% 
        ggplot(aes(x=PValue)) + 
        geom_histogram(breaks =seq(0,1,.1) ,col=1)
    )
    
for (i in 1:ncol(L)) 
    histsP[[i]] <- histsP[[i]] +
    ggtitle(paste("contrast",names(topTablesF)[i]))
histsP
```

TODO: some of these plots might seem a bit weird.

### Checks

```{r}
genes2.5h = topTablesF$treatmentVitD$table
genes4h = topTablesF$`treatmentVitD + time4h:treatmentVitD`$table
genes24h = topTablesF$`treatmentVitD + time24h:treatmentVitD + time4h:treatmentVitD`$table
```

TODO: actual interpretation

#### Percentages upregulated vs downregulated genes

In the paper:
- 2.5h: 100% up
- 4h: 59% up
- 24h: 55.1% up

TODO: why the mismatch?

```{r}
sig2.5 = genes2.5h %>% filter(FDR < 0.05)
c2.5up = sig2.5 %>% filter(logFC > 0) %>% count
c2.5 = sig2.5 %>% count

print(c2.5up / c2.5)

sig4 = genes4h %>% filter(FDR < 0.05)

c4up = sig4 %>% filter(logFC > 0) %>% count - c2.5up
c4 = sig4 %>% count - c2.5

print(c4)
print(c4up / c4)


sig24 = genes24h %>% filter(FDR < 0.05)
c24up = sig24 %>% filter(logFC > 0) %>% count - c4up - c2.5up
c24 = sig24 %>% count - c4 - c2.5

print(c24)
print(c24up / c24)

```

#### Number of genes that only become significant after 24h
In the paper they mention that only after 24h of treatment, 995 genes become significantly expressed.

We count:

```{r}
at_24h = sig_counts["treatmentVitD + time24h:treatmentVitD + time4h:treatmentVitD"]
at_4h = sig_counts["treatmentVitD + time4h:treatmentVitD"]

at_24h - at_4h
```

#### NOD2

According to this paper (*Epigenome-wide effects of vitamin D and their impact on the transcriptome of human monocytes involve CTCF*),
expression of this gene ENSG00000167207 (NOD2), is only affected after 24h .

Evaluate this.

```{r}
gene = "ENSG00000167207"

genes2.5h[gene,]$FDR
genes4h[gene,]$FDR
genes24h[gene, ]$FDR
```
Indeed, only significant after 24h of treatment.

#### HTT
HTT: ENSG00000197386

Should become significant after 4h of treatment.

```{r}
gene = "ENSG00000197386"



genes2.5h[gene,]$FDR
genes4h[gene,]$FDR
genes24h[gene, ]$FDR
```
Indeed, only significant after 4h of treatment.
