---
output:
  pdf_document: default
  html_document: default
---
# RNAseq Data - GSE69284

## General info
This dataset (GSE69284) was generated in 2016 and aimed to identify genes regulated by Vit D in the human monocytic THP-1 cell line. Expression profiling was performed through high throughput RNA sequencing after treatment with Vit D3 for 2.5 / 4 / 24 h. This study was performed on 18 samples (in triplicate form), 4 of which were used for our data analysis.

- Read length: 51bp
- Single/paired end sequencing: single end sequencing
- Platform used: Illumina NextSeq500

## Quality control
A MultiQC quality control was performed and mapping statistics are provided through the kallisto log output.

## Preprocessing RNA-seq data

### Trimming with Trimmomatic
Before mapping, the data was trimmed based on the quality scores:

```
java -jar $EBROOTTRIMMOMATIC/trimmomatic-0.39.jar SE -threads 8 -phred33 \
          $ID.fastq $ID\_trimQuality.fastq SLIDINGWINDOW:4:15
```

And to remove Illumina adapter sequences:

```
java -jar $EBROOTTRIMMOMATIC/trimmomatic-0.39.jar SE -threads 8 -phred33 \
          $ID\_trimQuality.fastq $ID\_trimBoth.fastq \
          ILLUMINACLIP:Adapters.fa:2:30:10
```

### Summerizing Gene Counts With Kallisto

Read mapping was performed using the Kallisto pseudoalignment tool.
This tools returns RNA-seq transcript counts, these results will be summerized to gene-level counts for the downstream analysis.

The human reference GRCh38.p13 (release 108) from Ensembl was used to create an index.
The following command was used to invoke Kallisto:

```
kallisto quant -i Homo_sapiens_index -o $ID\_quant -t 8 \
                --single -l 100 -s 1\
                -g Homo_sapiens.GRCh38.108.gtf $ID\_trimBoth.fastqg
```

### Preprocessing Statistics
The statistics of this mapping run are the following:

```{r}
preprocess_statistics <- read.csv("stats/RNAseq/GSE69284/preprocess.tsv",
                                  sep="\t")
preprocess_statistics
```

## Downstream Analysis

### Loading Libraries

```{r echo=FALSE, results='hide', error=FALSE, warning=FALSE, message=FALSE}
# install packages if not yet installed.
pkgs <- c("tximport", "dplyr", "edgeR", "biomaRt", "limma")
notInstalled <- pkgs[!pkgs %in% installed.packages()[,1]]
if(length(notInstalled) > 0){
  BiocManager::install(notInstalled)
}

library(tximport)
library(dplyr)
library(edgeR)
library(biomaRt)
```

### Pseudo count table

Kallisto transcript mapping was performed to generate a pseudo-count table.
These pseudo-counts are transcript-level counts and need to be summerized into gene-level counts.

The transcript annotation data is extracted.

```{r}
# annotation data
GRCh38 <-  useEnsembl(biomart="genes", host="https://www.ensembl.org",
                      dataset="hsapiens_gene_ensembl")

# attributes
atr <- listAttributes(GRCh38)
data <- getBM(attributes = c('ensembl_gene_id', 'ensembl_transcript_id',
                             'external_gene_name'), mart = GRCh38)

tx2geneGtf <- dplyr::select(data, ensembl_transcript_id, ensembl_gene_id)

# maps transcript -> gene
tx2geneGtf <- rename(tx2geneGtf, TXNAME = ensembl_transcript_id)
tx2geneGtf <- rename(tx2geneGtf, GENEID = ensembl_gene_id)
head(tx2geneGtf)
```

The Kallisto pseudoalignment data is loaded.

Search for the Kallisto output files.

```{r}
files <- list.files("./stats/RNAseq/GSE69284", recursive=TRUE)

abundance_files <- grep("abundance_bis.tsv", files, value=TRUE)
samples <- gsub("(.*)_quant.*", "\\1", abundance_files, perl=TRUE)
abundance_files <- paste(rep("./stats/RNAseq/GSE69284/"),
                         abundance_files, sep="")

names(abundance_files) <- samples
```

Load RNAseq data using tximport.
This function performs gene-level summerization of the transcript counts.

```{r}
txi <- tximport(abundance_files, type = "kallisto", tx2gene = tx2geneGtf)
```

Check for duplicate gene counts

```{r}
dups = duplicated(rownames(txi$counts))
sum(dups)
```

As the sum is 0 there are no duplicate gene counts.

#### Annotation

General sample annotation file was provided.  
A case-control study was performed with 3 different treatment groups (2.5, 4, and 24 hour exposure) in the THP-1 cell line. The treatment time will be added as a factor during statistical analysis.

```{r}
RNAseq_annot <- read.csv("./data/annotation/GSE69284/SraRunTable.txt")
colnames(RNAseq_annot)

annotation <- RNAseq_annot[c("TREATMENT", "Treatment_Time")]
colnames(annotation) <- c("treatment", "time")
rownames(annotation) <- RNAseq_annot$Run

# convert into factors
annotation$treatment <- factor(annotation$treatment, 
                               levels=c("Ethanol", "100 nM 1\\,25(OH)2D3"),
                               labels=c("Ethanol", "VitD"))
annotation$time <- factor(annotation$time, levels=c("2.5 h", "4 h", "24 h"),
                          labels=c("2.5h", "4h", "24h"))

annotation
```

Check if the colnames match up in order with the annotation data.

```{r}
rownames(annotation) == colnames(txi$counts)
```

## EdgeR Analysis

### Normalisation

As we're working with *pseudocounts* (summerized gene counts from transcript counts), we need to add extra bias correction offsets (based on average transcript length) to make the data fit for EdgeR gene-level analysis.

```{r}
## Make tpm values compatible with edgeR
cts <- txi$counts
normMat <- txi$length
# Obtaining per-observation scaling factors for length, adjusted to avoid
# changing the magnitude of the counts.
normMat <- normMat/exp(rowMeans(log(normMat)))
normCts <- cts/normMat

# Computing effective library sizes from scaled counts, to account for
# composition biases between samples.
eff.lib <- calcNormFactors(normCts) * colSums(normCts)
# Combining effective library sizes with the length factors, and calculating
# offsets for a log-link GLM.
normMat <- sweep(normMat, 2, eff.lib, "*")
normMat <- log(normMat)
```

### Create DGEList object

```{r}
dge <- DGEList(cts)
dge <- scaleOffset(dge, normMat)
```

### Design Matrix

```{r}
design <- model.matrix(~time * treatment, annotation)
rownames(design) <- colnames(dge)
design
```

### Filtering

```{r}
keep <- filterByExpr(dge,design)
print(table(keep))
dge <- dge[keep, , keep.lib.sizes=FALSE]
```

### Data exploration

MDS plot

```{r}
limma::plotMDS(dge,labels=paste(annotation$treatment,annotation$time,sep=" - "), 
               col=as.double(annotation$treatment))
```

Observations:
- maybe no time effect for Ethanol treated samples? (expected)
- time effect for VitD treated samples (expected)


### Estimate dispersion

```{r}
dge <- estimateDisp(dge, design)  
plotBCV(dge)
```

Looks OK? -> TODO: why

### Fit quasi-negative binomial model


```{r}
fit <- glmQLFit(dge,design)
```

### Define contrasts

TODO: do the contrasts properly

```{r}
L <- msqrob2::makeContrast(
    c("treatmentVitD = 0",
      "time24h:treatmentVitD = 0",
      "time4h:treatmentVitD = 0",
      "treatmentVitD + time4h:treatmentVitD = 0",
      "treatmentVitD + time24h:treatmentVitD = 0",
      "time4h = 0",
      "time24h = 0"),
    parameterNames = colnames(design))
L
```

Perform Contrast Tests

```{r}
testsF <- apply(L, 2, function(fit,contrast) 
    glmQLFTest(fit,contrast=contrast), 
    fit = fit)

topTablesF <- lapply(testsF, topTags, n=nrow(dge))

sig_counts <- sapply(topTablesF, function(x) sum(x$table$FDR< 0.05))
sig_counts
```

`time4h` and `time24h` capture the contrasts comparing the time effect on non-treated cells, there shouldn't really be a time effect, which is what is seen by the fact that not a lot of significant genes were found for these contrasts.

Similar to the conclusions on the original paper, there is a considerate time effect on the treatment,
in the contrast ` there is a lot of expressed genes compared to `treatmentVitD` which only captures the contrast comparing 2.5h treated cells vs 2.5h non-treated cells.

### Volcano Plots

```{r}
library(ggplot2)
for (i in 1:ncol(L))
{
   volcano<- ggplot(topTablesF[[i]]$table, 
                    aes(x=logFC,y=-log10(PValue),color=FDR < 0.05)) 
   + geom_point() + scale_color_manual(values=c("black","red")) 
   + ggtitle(paste("contrast",names(topTablesF)[i]))
print(volcano)
}
```

### Histograms of P-values

```{r}
histsP <- lapply(topTablesF, function(x) 
    x$table %>% 
        ggplot(aes(x=PValue)) + 
        geom_histogram(breaks =seq(0,1,.1) ,col=1)
    )
    
for (i in 1:ncol(L)) 
    histsP[[i]] <- histsP[[i]] +
    ggtitle(paste("contrast",names(topTablesF)[i]))
histsP
```

### Save Results

```{r}
genes2.5h <- topTablesF$treatmentVitD$table
genes4h <- topTablesF$`treatmentVitD + time4h:treatmentVitD`$table
genes24h <- topTablesF$`treatmentVitD + time24h:treatmentVitD`$table
```

### Checks

Convert gene symbol into ensembl id.

```{r}
symbol2ensembl = function(x) {
  return(data[data$external_gene_name == x, "ensembl_gene_id"][1])
}
```


#### Paper result: time effect of treatment
In the paper they mention a few genes that only express after a certain treatment length.


#### NOD2

According to this paper (*Epigenome-wide effects of vitamin D and their impact on the transcriptome of human monocytes involve CTCF*),
expression of this gene (NOD2), is only affected after 24h.

Evaluate this.

```{r}
gene <- symbol2ensembl("NOD2")

genes2.5h[gene,]$FDR
genes4h[gene,]$FDR
genes24h[gene, ]$FDR
```

Indeed, only somewhat significant after 24h of treatment.

#### HTT

Should become significant after 4h of treatment according to the paper results.

```{r}
gene <- symbol2ensembl("HTT")

genes2.5h[gene,]$FDR
genes4h[gene,]$FDR
genes24h[gene, ]$FDR
```

Indeed, only somewhat significant after 4h of treatment.

### Gene Set Analysis

We will use Kegga to gather gene ontology terms.

Need to incorporate transcript length in the analysis as it is easier for longer transcripts to be expressed as it will have more reads, and thus higher counts (which also means lower variance and thus easier to detect as differentially expressed).

```{r}
library(limma)

# to map the ensembl gene ids to entrez ids (input for kegga)
gid <- getBM(attributes = c("ensembl_gene_id", "external_gene_name", 
                            "entrezgene_id", "transcript_length"), mart=GRCh38)

gid <- gid[!duplicated(gid$ensembl_gene_id),]
```

Universe (expressed genes after filtering)

```{r}
univ <- genes4h
gid_f = gid[gid$ensembl_gene_id %in% rownames(univ), ]
gid_f_sort <- gid_f[order(gid_f$ensembl_gene_id),]

univ$gene_symbol <- gid_f_sort$external_gene_name
univ$entrez_gene_id <- gid_f_sort$entrezgene_id

univ <- univ[!is.na(univ$entrez_gene_id),]
```

```{r}
sig2.5 <- genes2.5h[genes2.5h$FDR < 0.5,]
print(dim(sig2.5))

sig4 <- genes4h[genes4h$FDR < 0.5,]
sig4 <- sig4[!(rownames(sig4) %in% rownames(sig2.5)), ]
print(dim(sig4))


sig24 <- genes24h[genes24h$FDR < 0.5,]
sig24 <- sig24[!(rownames(sig24) %in% rownames(sig4) 
                 | rownames(sig24) %in% rownames(sig2.5)), ]
```

```{r}
for (idx in 1:3) {

    if (idx == 1) {
        genes <- sig2.5
    } else if (idx == 2) {
        genes <- sig4
    } else {
        genes <- sig24
    }      

    # relax as otherwise too less significant genes
    # filter out the DE genes
    print(dim(genes))

    gid_f = gid[gid$ensembl_gene_id %in% rownames(genes), ]
    genes_sort <- genes[order(rownames(genes)),]
    gid_f_sort <- gid_f[order(gid_f$ensembl_gene_id),]

    genes_sort$gene_symbol <- gid_f_sort$external_gene_name
    genes_sort$entrez_gene_id <- gid_f_sort$entrezgene_id
    genes_sort$transcript_length <- gid_f_sort$transcript_length

    # Analysis with Kegga:

    # not all genes have an entrez idz
    genes_sort <- genes_sort[!is.na(genes_sort$entrez_gene_id),]

    # filter out NA values

    res <- kegga(de=genes_sort$entrez_gene_id,
                 universe = univ$entrez_gene_id,
                 species.KEGG="hsa", trend=genes_sort$transcript_length)
    res <- res[order(res$N),]
    
    res$FDR.DE <- p.adjust(res$P.DE, n=nrow(res), method="BH")
    print(head(res[order(res$DE, decreasing=TRUE),]))
}
```
