---
title: "VitD-OmicsAnalysis"
author: "Robbe & Simon"
date: "`r Sys.Date()`"
output: html_document
---

# Introduction
## Project Overview
Group number: 11
Group memebers: Robbe Claeys & Simon Bil  

**Research question:** An omics-wide analysis of Vitamin D difieciency.

## Data Overview
**Paper links:**

ChIP seq

- https://pubmed.ncbi.nlm.nih.gov/25667505/
- https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2945184/#ack-1title

ChIP & RNA seq

- https://www.nature.com/articles/s41598-021-86032-5

RNA microarray

- https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4030167/  

**Download links:**

ChIP seq

- https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE22484
- https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE89431

RNA seq

- https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE69303

RNA microarray

- https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE5145

```{r}
table_general_info <- rbind(c("GSE5145", "GSE89431", "GSE89431", "GSE69303"), c(6, 16, 39, 2), c(6, 16, 6, 2))

colnames(table_general_info) <- c("Microarray Data", "ChIPseq Data", "ChIPseq Data", "RNAseq Data")

rownames(table_general_info) <- c("Accesion ID", "Number of Samples (original)", "Number of Samples (analysed)")

table_general_info
```
# Microarray Data
## General info
This dataset (GSE5145) was generated in 2007 and aimed to identify the genes regulated by Vit D in primary bronchial smooth muscle cells. Cells were treated for 24 hours with 100 nM of Vit D3. The experiment was done in triplicates (a total of 6 samples were analyzed). All samples of this study were used for our data analysis.

**Microarray type:** Affymetrix Human Genome U133 Plus 2.0 Array [HG-U133_Plus_2]

## Intensity values
Load libraries
```{r}
library(affy)
library(affycoretools)
library(oligo)
library(arrayQualityMetrics)
library(ArrayExpress)
library(limma)
library(siggenes)
library(hgu133plus2.db)
```

Load the data
```{r}
id <- "GSE5145"
exonCELs <- list.celfiles("./data/microarray/")
affyRaw <- read.celfiles(paste(rep("./data/microarray/", length(exonCELs)), exonCELs, sep = ""))
```

Get dimensions and general overview of the data. The first three `.CEL` files are the Vit D deficient samples, the last three are the control samples.
```{r}
dim(exprs(affyRaw))
head(exprs(affyRaw))
```

## Annotation
We will be investigating differential expression between colon Vit D deficient and normal primary bronchial smooth muscle cell samples. No other confounding factors are present in the sample annotation (so far).

The annotation was provided through a `.soft` file (https://warwick.ac.uk/fac/sci/moac/people/students/peter_cock/r/geo/)

# ChIPseq Data

# RNAseq Data

