---
title: "Group 11 - VitD-OmicsAnalysis"
author: "Robbe & Simon"
date: "`r Sys.Date()`"
output: html_document
---

# Introduction
## Project Overview
Group number: 11
Group memebers: Robbe Claeys & Simon Bil  

**Research question:** An omics-wide analysis of Vitamin D difieciency.

## Data Overview
**Paper links:**

ChIP seq

- https://pubmed.ncbi.nlm.nih.gov/25667505/
- https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2945184/#ack-1title

ChIP & RNA seq

- https://www.nature.com/articles/s41598-021-86032-5

RNA microarray

- https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4030167/  

**Download links:**

RNA microarray

- https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-GEOD-22523?query=vitamin%20D

RNA seq

- https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE69284

ChIP seq

- https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE22484
- https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE89431

```{r}
table_general_info <- rbind(c("E-GEOD-22523", "GSE69284", "GSE89431", "GSE89431"), c(8, 18, 39, 2), c(8, 6, 6, 2))

colnames(table_general_info) <- c("Microarray Data", "RNAseq Data", "ChIPseq Data", "ChIPseq Data")

rownames(table_general_info) <- c("Accesion ID", "Number of Samples (original)", "Number of Samples (analysed)")

table_general_info
```

# Microarray Data - GSE22523
## General info
This dataset (E-GEOD-22523) was generated in 2010 and aimed to identify genes regulated by Vit D in B cells. A gene expression microarray analysis in HVDRR EBV-transformed B-cells and control cells was performed in the presence or absence of vitamin D. This study was performed on 8 samples (4 cases & 4 controls), all were used for our data analysis.

**Microarray type:** Affymetrix Human Gene 1.0 ST Array [HuGene-1_0-st]

## Intensity values
Load libraries
```{r echo = T, results = 'hide', error=FALSE, warning=FALSE, message=FALSE}
library(affy)
library(oligo)
library(arrayQualityMetrics)
library(ArrayExpress)
library(limma)
library(siggenes)
```

Load the data
```{r echo = T, results = 'hide', error=FALSE, warning=FALSE, message=FALSE}
id <- "E-GEOD-22523"
exonCELs <- list.celfiles("./data/microarray/")
arrayRaw <- read.celfiles(paste(rep("./data/microarray/", length(exonCELs)), exonCELs, sep = ""))
```

```{r}
sets = ArrayExpress(E-GEOD-22523
```


The first 4 `.CEL` files contain the data for the hereditary vitamin D-resistant rickets (HVDRR) cells, of which the first two are biological replicates and the last two were treated with Vit D. The last 4 `.CEL` files contain the data for the Wild-type (WT) cell, again, of which the first two are biological replicates and the last two were treated with Vit D.

Get dimensions and general overview of the data. 
```{r}
dim(exprs(arrayRaw))
head(exprs(arrayRaw))
```

## Annotation
+ add fact about genetic background of the cells (https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-GEOD-22523?query=vitamin%20D) under factors table
We will be investigating differential expression between . No other confounding factors are present in the sample annotation (so far).

```{r}
sdrf <- read.delim("./data/annotation/GSE22523/E-GEOD-22523.sdrf.txt")
print(sdrf[, c("Source.Name", "Comment..Sample_source_name.", "Array.Design.REF")])
```

# ChIPseq Data - GSE89431

## General info
This dataset (GSE89431) was generated in 2017 and aimed to identify genomic locations bound by the vitamin D receptor (VDR).
ChIP-seq for the vitamin D receptor (VDR) was performed in THP-1 cells after stimulation with the natural VDR ligand 1,25-dihydroxyvitamin D3 (1,25(OH)2D3) or with vehicle for 2 or 24h before ChIP assay.

- Read length: 51bp
- Single/paired end sequencing: single end sequencing
- Platform used: Illumina HiSeq 2000

## Quality control on fastQ files

```{r, echo=FALSE, results='asis'}
htmltools::includeHTML("./stats/ChIPseq/GSE89431/multiqc_report.html")
```

- Watch out bcause there are two samples with lots of reads (fastQC sequence counts)
- time as a factor in the model -> discrete, not continuous

## Mapping statistics

> We tried to perform mapping ourselves, but failed at a certain point.
> For this reason we did not have time to finish peak data and annotation on the ChIPseq data.

## Annotation

```{r}
gse89431_annot <- read.csv("./data/annotation/GSE89431/SraRunTable.txt")
print(gse89431_annot[c("Run", "Treatment", "Time")])
```

# ChIPseq Data - GSE22484
## General info
This dataset (GSE22484) was generated in 2010 and aimed to identify the genomic locations bound by the vitamin D receptor (VDR) in two lymphoblastoid cell lines before and after calcitriol (active form of vitamin D) treatment for 36 hours.

- Read length: 36bp
- Single/paired end sequencing: single end sequencing
- Platform used: Illumina Genome Analyzer II (Homo sapiens)

## Quality control

```{r, echo=FALSE, results='asis'}
htmltools::includeHTML("./stats/ChIPseq/GSE22484/multiqc_report.html")
```

## Mapping statistics

> We tried to perform mapping ourselves, but failed at a certain point.
> For this reason we did not have time to finish peak data and annotation on the ChIPseq data.

## Annotation

```{r,}
gse22484_annot <- read.csv("./data/annotation/GSE22484/SraRunTable.txt")
print(gse22484_annot[c("Run", "Treatment", "SAMPLE_TYPE")])
```

# RNAseq Data - GSE69284
## General info
This dataset (GSE69284) was generated in 2016 and aimed to identify genes regulated by Vit D in the human monocytic THP-1 cell line. Expression profiling was performed through high throughput RNA sequencing after treatment with Vit D3 for 2.5 / 4 / 24 h. This study was performed on 18 samples (in triplicate form), 6 were used for our data analysis (1 replicate at all 3 treatments).

Better to use all samples (all 18 and add time as factor)

- Read length: 51bp
- Single/paired end sequencing: single end sequencing
- Platform used: Illumina NextSeq500

## Quality control
A MultiQC quality control was performed and mapping statistics are provided through the kallisto log output.
```{r, echo = FALSE, results='asis'}
htmltools::includeHTML("./stats/RNAseq/multiqc_report.html")
```

## Pseudo count table
```{r}
library(tximport)

# get file locations
files <- list.files("./stats/RNAseq/kallisto")
files <- files[grep("abundance.tsv", files)]
samples <- unlist(strsplit(files,"_"))[c(1:length(files))*2-1]
files <- paste(rep("./stats/RNAseq/kallisto",length(files)), files, sep="")
names(files) <- samples

# load RNAseq data
#txi <- tximport(files, type = "kallisto", tx2gene = tx2geneGtf)
```


## Annotation
General sample annotation file was provided.  
A case-control study was performed with 3 different treatment groups (2.5, 4, and 24 hour exposure) in the THP-1 cell line.
```{r}
RNAseq_annot <- read.csv("./data/annotation/GSE69284/SraRunTable.txt")
print(RNAseq_annot[c("Run", "Treatment", "treatment_time")])
```
