---
title: "Group 11 - VitD-OmicsAnalysis"
author: "Robbe & Simon"
date: "`r Sys.Date()`"
output: html_document
---

# Introduction
## Project Overview
Group number: 11
Group members: Robbe Claeys & Simon Bil  

**Research question:** An omics-wide analysis of Vitamin D difieciency.

## Data Overview
**Paper links:**

RNA microarray

- https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4030167/
  - https://doi.org/10.1152/physiolgenomics.00134.2006
  - https://doi.org/10.1096/fj.10-172577

ChIP & RNA seq

- https://www.nature.com/articles/s41598-021-86032-5

**Download links:**

RNA microarray (GSE5145)

- https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE5145

RNA microarray (GSE22523)

- https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE22523

RNA seq (E-MTAB-9789)

- https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-9789?query=%22vitamin%20d%22%20

RNA seq (GSE69284)

- https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE69284

ChIP seq (GSE27438)

- https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE27438

```{r}
table_general_info <- rbind(c("GSE5145", "GSE22523", "E-MTAB-9789", "GSE69284", "GSE89431"), c(6, 8, 112, 18, 39), c(6, 8, 32, 4, 6))

colnames(table_general_info) <- c("Microarray Data", "Microarray Data", "RNAseq Data", "RNAseq Data", "ChIPseq Data")

rownames(table_general_info) <- c("Accesion ID", "Number of Samples (original)", "Number of Samples (analysed)")

table_general_info
```

---

Loading all libraries

```{r echo = False, results = 'hide', error=FALSE, warning=FALSE, message=FALSE}
library(htmltools)

# for microarrays
library(oligo)
library(GEOquery)
library(arrayQualityMetrics)
library(limma)

# for RNA seq
library(tximport)
library(dplyr)
library(edgeR)

library(ggplot2)

```

# Microarray Data

## General info

This dataset was generated in 2007 and aimed to identify the genes regulated by Vit D in primary bronchial smooth muscle cells. Cells were treated for 24 hours with 100 nM of Vit D3. The experiment was done in triplicate form (a total of 6 samples were analyzed). All samples of this study were used for our data analysis.

**Microarray type:** Affymetrix Human Genome U133 Plus 2.0 Array [HG-U133_Plus_2]

## Download & Extract the data

```{r eval=FALSE}
url = "https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE5145&format=file"
tf = tempfile()
download.file(url, tf)
untar(tarfile=tf, exdir="./data/microarray/GSE5145")
unlink(tf)
```

Download annotation file.

```{r eval=FALSE}
url = "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE5nnn/GSE5145/soft/GSE5145_family.soft.gz"
download.file(url, "./data/microarray/GSE5145/GSE5145_family.soft.gz")
```

Extract the .gz files.

```{r eval=FALSE}
l <- list.files("./data/microarray/GSE5145", pattern=".+\\.gz",  full.names = TRUE)
print(l)
lapply(l, function(x) { gunzip(x, overwrite=TRUE) } )
```

## Intensity values

Load the data.

```{r echo = T, results = 'hide', error=FALSE, warning=FALSE, message=FALSE}
exonCELs <- list.celfiles("./data/microarray/GSE5145/")
arrayRaw <- read.celfiles(paste(rep("./data/microarray/GSE5145/", length(exonCELs)), exonCELs, sep = ""))
```

Get dimensions and general overview of the data. The first three `.CEL` files are the Vit D deficient samples, the last three are the control samples.

```{r}
dim(exprs(arrayRaw))
head(exprs(arrayRaw))
```

Extraction of the expression values using the robust multichip average method (RMA) and summarization of the array data.

```{r}
arrayRma = rma(arrayRaw)
dim(exprs(arrayRma))
head(exprs(arrayRma))
```

## Annotation

We will be investigating differential expression between Vit D deficient and normal primary bronchial smooth muscle cell samples. No other confounding factors are present in the sample annotation. All primary bronchial smooth muscle cells were derived from the same tissue sample of a 31 year old Hispanic male.

```{r error=FALSE, warning=FALSE, message=FALSE}
annot <- getGEO(filename = "./data/microarray/GSE5145/GSE5145_family.soft")

IDs <- annot@header$sample_id
titles <- c()

for (id in IDs) {
  title <- annot@gsms[[id]]@header[["title"]]
  titles <- append(titles, title)
}
data.frame(IDs, titles)
```

## Explorative quality evaluation

Quality of the array determined with the arrayQualityMetrics package.

```{r}
arrayQualityMetrics(expressionset = arrayRma,
                    outdir = "./stats/Microarray/GSE5145/",
                    force = TRUE)
```

From this analysis we can conclude that the overall quality of the arrays is quite good. However, sample 6 shows signs of being an outlier and the graph on figure 7 (Standard deviation versus rank of the mean) shows an upward trend, which is symptomatic of a saturation of the intensities.

## Differential expression analysis

We will perform LIMMA to look for differential expressed genes.

```{r}
# design matrix
design <- model.matrix(~ 0 + factor(c(1,1,1,2,2,2)))
colnames(design) <- c("VitD", "Control")
fit <- lmFit(arrayRma, design)
```

To make a pair-wise comparison between the two groups the appropriate contrast matrix is created.

```{r}
contrast.matrix <- makeContrasts(Control-VitD, levels = design)
fit2 <- contrasts.fit(fit, contrast.matrix)
fit2 <- eBayes(fit2)
```

A list of top genes differential expressed in group2 (Control) versus group1 (VitD) can now be obtained.

The best probe (203887_s_at) is expressed 10^3 times less in the Control group (log fold change). All 500 probes in this list have significant adjusted p-values.

```{r}
topNumber <- 500
table <- topTable(fit2, coef=1, number=topNumber, adjust="BH") # FDR correction of p-values
table
```

## Summarization at the gene level

Note that we are now working at the level of probe sets and not at level of genes per se. Whereas many genes are only represented by a single probe set on a micro array, for some genes multiple probe sets are present

```{r}
# input data (all of equal length)
allProbeIDs <- annot@gpls[["GPL570"]]@dataTable@table[["ID"]]
allGeneIDs <- annot@gpls[["GPL570"]]@dataTable@table[["ENTREZ_GENE_ID"]]
allGeneSym <- annot@gpls[["GPL570"]]@dataTable@table[["Gene Symbol"]]
allGeneOnt <- annot@gpls[["GPL570"]]@dataTable@table[["Gene Ontology Biological Process"]]
```

```{r warning=FALSE}
# desired data
geneIDs <- c()
geneSym <- c()
logFCs <- c()
geneOnt <- c()

for (topIndex in 1:topNumber) {
  probeID <- rownames(table)[topIndex]
  annotIndex <- grep(probeID, allProbeIDs)
  
  if (!(allGeneIDs[annotIndex] %in% geneIDs) && (probeID %in% allProbeIDs)) {
    geneIDs <- append(geneIDs, allGeneIDs[annotIndex[1]])
    geneSym <- append(geneSym, allGeneSym[annotIndex[1]])
    logFCs <- append(logFCs, table[[1]][topIndex])
    geneOnt <- append(geneOnt, allGeneOnt[annotIndex[1]])
  } 
}

df <- data.frame(Entrez.gene.ID = geneIDs, Gene.symbol = geneSym, Fold.Change = logFCs, Gene.ontology = geneOnt)
df
```

The top 500 most significant probes could be mapped to 330 individual genes, which were mostly upregulated after Vit D treatment. Many of these genes have a role in high-level signal transduction (cell cyle, metabolism, differentiation). This is a first indication that Vit D is an important nutrient with pleiotropic regulatory effects in the cell.

When comparing these findings to those reported in the original paper, the same upregulated genes were found.

---

# Microarray Data - GSE22523
## General info
This dataset (E-GEOD-22523) was generated in 2010 and aimed to identify genes regulated by Vit D in B cells. A gene expression microarray analysis in HVDRR EBV-transformed B-cells and control cells was performed in the presence or absence of vitamin D. This study was performed on 8 samples (4 cases & 4 controls), all were used for our data analysis.

**Microarray type:** Affymetrix Human Gene 1.0 ST Array [HuGene-1_0-st]

## Intensity values
Load the data
```{r echo = T, results = 'hide', error=FALSE, warning=FALSE, message=FALSE}
exonCELs <- list.celfiles("./data/microarray/GSE22523/")
arrayRaw <- read.celfiles(paste(rep("./data/microarray/GSE22523/", length(exonCELs)), exonCELs, sep = ""))
```

The first 4 `.CEL` files contain the data for the hereditary vitamin D-resistant rickets (HVDRR) cells, of which the first two are biological replicates and the last two were treated with Vit D. The last 4 `.CEL` files contain the data for the Wild-type (WT) cell, again, of which the first two are biological replicates and the last two were treated with Vit D.

Get dimensions and general overview of the data. 
```{r}
dim(exprs(arrayRaw))
head(exprs(arrayRaw))
```

Summarization of the array data.
```{r}
arrayRma = rma(arrayRaw)
dim(exprs(arrayRma))
head(exprs(arrayRma))
```

## Annotation
We will be investigating differential expression between VitD treated & untreated cel cultures. Confounding factors: there are 4 wild-type and 4 HVDRR cells.

```{r error=FALSE, warning=FALSE, message=FALSE}
annot <- getGEO(filename = "./data/annotation/GSE22523/GSE22523_family.soft")

IDs <- annot@header$sample_id
titles <- c()

for (id in IDs) {
  title <- annot@gsms[[id]]@header[["title"]]
  titles <- append(titles, title)
}
data.frame(IDs, titles)
```

## Explorative quality evaluation

Quality of the array determined with the arrayQualityMetrics package.

```{r warning=FALSE}
arrayQualityMetrics(expressionset = arrayRma,
                    outdir = "./stats/Microarray/GSE22523/",
                    force = TRUE)
```

From this analysis we can conclude that the overall quality of the arrays is very good. No problematic outliers or variance trends were observed.

## Differential expression analysis

We will perform LIMMA to look for differential expressed genes.

```{r}
design <- model.matrix(~ factor(c(1,1,2,2,1,1,2,2)) + 
    factor(c(1,1,1,1,2,2,2,2)))
colnames(design) <- c("Intercept", "Treated", "Celltype")
fit <- lmFit(arrayRma, design)
fit <- eBayes(fit)
```

A list of top differential expressed genes in the treated groups versus the control groups can now be obtained by evaluating the first coefficient. The cell type coefficient did not yield significant findings, indicating that cell type did not play a deciding role in the differential expression.

The best probe (7894912) is expressed 10^-0.9 times less in the Control group (log fold change). Notably, no significant findings were retrieved after FDR correction of the p-values.

```{r}
topNumber <- 500
table <- topTable(fit, coef = 2, number = topNumber, adjust = "BH") # FDR correction of p-values
table
```

## Summarization at the gene level

For annotation, the different probe sets are mapped to individual genes present in the .soft file.

```{r}
# input data (all of equal length)
allProbeIDs <- annot@gpls[["GPL6244"]]@dataTable@table[["ID"]]
allGeneIDs <- annot@gpls[["GPL6244"]]@dataTable@table[["gene_assignment"]]
```

```{r}
# desired data
geneIDs <- c()
logFCs <- c()

for (topIndex in 1:topNumber) {
  probeID <- rownames(table)[topIndex]
  annotIndex <- grep(probeID, allProbeIDs)
  
  if (!(allGeneIDs[annotIndex] %in% geneIDs) & (probeID %in% allProbeIDs)) {
    geneIDs <- append(geneIDs, allGeneIDs[annotIndex[1]])
    logFCs <- append(logFCs, table[[1]][topIndex])
  } 
}

df <- data.frame(Entrez.gene.ID = geneIDs, Fold.Change = logFCs)
df
```

The top 500 most significant probes could be mapped to 235 individual genes. Again, we see many general signal transduction factors (e.g. elongation factor) as well as general-purpose factors such as the mitochondrial rRNA transcripts.

When comparing these findings to those reported in the original paper, only a few matches were found. This is because the paper was specifically investigating transcripts with a role in immune signaling, thus only these were evaluated.

---

TODO: include other file?

# RNAseq Data - E-MTAB-9789

## General info
This dataset was generated in 2021 and aimed to identify genes regulated by Vit D in saline-injected skin of healthy old adults through RNA sequence analysis before and after Vit D3 supplementation. The study wanted to verify whether Vitamin D insufficiency may exacerbate non-specific inflammation observed in older adults. This data set originally contained 112 samples from 57 individuals, this was reduced to 32 samples from 16 individuals for our analysis.

Due to the HPC downtime, pre-generated count tables were used as a starting point instead of the raw sequence output. RNAseq data were mapped to the reference transcriptome (Ensembl Human GRCh38 release 99) using Kallisto. Transcript-level counts were summed on gene level using tximport.

- Read length: 41bp
- Single/paired end sequencing: paired end sequencing
- Platform used: Illumina NextSeq500

## Download the data

## Loading count data
The count data was downloaded and 32 samples (from 16 individuals) were selected.

```{r}
# loading the annotation file
annot <- read.table("./data/RNAseq2/E-MTAB-9789.sdrf.txt", header = TRUE, sep = "\t")
# removing duplicate entries (replicate correction has already been performed)
annot <- annot[!duplicated(annot[ , "Derived.Array.Data.File"]),]
# sort based on individual
annot <- annot[order(annot$Characteristics.individual.), ]

print(annot[c("Source.Name", "Characteristics.individual.", "Factor.Value.compound.")])

# select and factorize variables we wish to incorporate in the model
individuals <- factor(annot$Characteristics.individual.)
treatments <- factor(annot$Factor.Value.compound.)
treatments <- relevel(treatments, ref = "none")
```


```{r}
# first loading the annotation file
annot <- read.csv("./data/RNAseq/E-MTAB-9789.csv", header = TRUE, sep = ",")
head(annot)

# select and factorize variables we wish to incorporate in the model
individuals <- factor(annot$Characteristics.individual.)
treatments <- factor(annot$Factor.Value.compound.)
treatments <- relevel(treatments, ref = "none")
```

Gender, age, and the measured Vit D serum levels (before and after treatment) are possible confounding factors that can be added to the model. We chose to only correct for individual in order to reduce complexity.

```{r}
# loading the counts & generating a table
df <- data.frame(matrix(nrow = 23402))
  
for (row in 1:nrow(annot)) {
  # get required entry information for generating count table
  indiv <- annot[row,]$Characteristics.individual.
  treatment <- annot[row,]$Factor.Value.compound.
  filePath <- paste("./data/RNAseq2/txts/", annot[row,]$Derived.Array.Data.File, sep = "")
  if (treatment == "none") {
    treatment <- "0"
  } else {
    treatment <- "1"
  }
  colName <- paste(row, treatment, indiv, sep = "_")
  counts <- read.table(filePath, sep = "\t", header = FALSE)$V2
  df[,colName] <- counts
}

# name rownames (= gene names)
rowNames <- read.table("./data/RNAseq2/txts/AD006S6_S5.txt", sep = "\t", header = FALSE)$V1
rownames(df) <- rowNames
# remove temporary nrow column
df <- subset(df, select = - matrix.nrow...23402.)
head(df)
```

A DGEList (edgeR count object) is created and TMM normalization is applied to correct for the library size for highly abundant differentialy expressed genes:

```{r}
y = DGEList(counts = df, group = treatments)
y = calcNormFactors(y)
y$samples
```

## Data exploration
The plotMDS(.) function draws a multi-dimensional scaling plot of the RNA samples in which distances correspond to leading log-fold-changes between each pair of RNA samples. The leading log-fold-change is the average (root-mean-square) of the largest absolute log-fold- changes between each pair of samples. It is equivalent to a principal component analysis, but scaled differently.

```{r}
plotMDS(y, col = as.numeric(treatments))
```

Distances on an MDS plot of a DGEList object correspond to leading log-fold-change between each pair of samples. Leading log-fold-change is the root-mean-square average of the largest log2-fold-changes between each pair of samples. Each pair of samples extracted from each patient tend to cluster together, suggesting a batch effect.

## The design matrix
Before we fit GLMs, we need to define our design matrix based on the experimental design. We want to test for differential expressions between Vit D treated and non-treated samples within batches, i.e. adjusting for differences between patients. In statistical terms, this is an additive linear model. So the design matrix is created as:

```{r}
# Create a design matrix that includes the treatments and patients
design <- model.matrix(~ treatments + individuals)
#design <- model.matrix(~ treatments)
rownames(design) <- colnames(df)
```

## Filter and normalize count data
Check for and remove duplicate rows.

```{r}
sum(duplicated(rownames(df)))
```

There is no point in analyzing genes that are not expressed in either experimental condition. We consider a gene to be expressed at a reasonable level in a sample if it has at least two counts for each million mapped reads in that sample. We also only keep genes expressed in at least three samples:

```{r}
#keep = rowSums(cpm(df) > 2) >= 3
keep <- filterByExpr(y)
df = df[keep, ]
table(keep)
```

```{r}
# more filtering
keep <- filterByExpr(y, design)
table(keep)
y <- y[keep, , keep.lib.sizes = FALSE]
```

```{r}
# more normalizing
y <- calcNormFactors(y)
y$samples
```

```{r}
# more exploration
plotMDS(y, col = as.numeric(treatments))
```

## Estimating the dispersion

The negative binomial distribution, is characterized by a dispersion parameter. The average dispersion over all genes is estimated:

```{r}
y <- estimateDisp(y, design)
plotBCV(y)
```

The genewise dispersions show a slight decreasing trend with expression level.

## Differential expression
To determine differentially expressed genes, genewise GLMs are fitted:

```{r}
# Fit a generalized linear model (GLM) using the design matrix
fit <- glmQLFit(y, design)

# Perform likelihood ratio tests to identify differentially expressed genes (coef = 2 => treated vs non-treated)
lrt <- glmQLFTest(fit, coef = 2)

# View the results of the likelihood ratio tests
result <- topTags(lrt, n = nrow(y))
result
resultsig <- result[result[["table"]][["FDR"]] < 0.05,]
resultsig

lrt$table %>% 
      ggplot(aes(x = PValue)) + 
      geom_histogram(breaks = seq(0, 1, .1) , col = 1)
```

After FDR-correction, 239 significant differentially expressed genes could be determined.

We check whether there was a genuine need to adjust for the experimental individuals. We do this by testing for differential expression between the individuals. There is considerable differential expression, justifying our decision to adjust for the batch effect:

```{r}
lrt = glmLRT(fit, coef = 3:19)
topTags(lrt)
```


---

# RNAseq Data - GSE69284

TODO: link separate file

## Annotation
General sample annotation file was provided.  
A case-control study was performed with 3 different treatment groups (2.5, 4, and 24 hour exposure) in the THP-1 cell line. The treatment time will be added as a factor during statistical analysis.
```{r}
RNAseq_annot <- read.csv("./data/annotation/GSE69284/SraRunTable.txt")
print(RNAseq_annot[c("Run", "TREATMENT", "Treatment_Time")])
```

---

# ChIPseq Data - GSE27438

## General info
This dataset (GSE27438) was generated in 2011 and aimed to identify genomic locations bound by the vitamin D receptor (VDR).
ChIP-seq for the vitamin D receptor (VDR) was performed in the monocytic THP-1 cells after stimulation with the natural VDR ligand 1,25-dihydroxyvitamin D3 (1,25(OH)2D3) or with vehicle for 4h before ChIP assay.

- Read length: 36bp
- Single/paired end sequencing: single end sequencing
- Platform used: Illumina Genome Analyzer II

## Quality control on fastQ files

```{r, echo=FALSE, results='asis'}
htmltools::includeHTML("./stats/ChIPseq/multiqc_report.html")
```

- Watch out bcause there are two samples with lots of reads (fastQC sequence counts)

## Mapping statistics

> We tried to perform mapping ourselves, but failed at a certain point.
> For this reason we did not have time to finish peak data and annotation on the ChIPseq data.

## Annotation

```{r}
gse27438_annot <- read.csv("./data/annotation/GSE27438/SraRunTable.txt")
print(gse27438_annot[c("Run", "Treatment")])
```

