#!/usr/bin/env bash

#PBS -l nodes=1:ppn=8
#PBS -l walltime=01:00:00
#PBS -l mem=12gb

# "# command" was used as a placeholder where a line of code should be inserted

# Load the necessary modules
module load FastQC/0.11.9-Java-11
module load parallel-fastq-dump/0.6.6-GCCcore-9.3.0-SRA-Toolkit-3.0.0-Python-3.8.2
module load MultiQC/1.9-intel-2020a-Python-3.8.2

# change current directory
cd $VSC_DATA_VO/Courses/AHTA/Group11/ChIPseq

# Make a directory to store fastqc output
mkdir 1_fastqc

# For loop over samples used (replicates A only)
for ID in SRR111951 SRR111952 SRR111953 SRR111954 SRR111955 SRR111956 SRR111957
do 
	# Echo the ID you are processing
	echo $ID
	
	# Load the reads for the ID (--sra-id $ID) you are processing using 8 cores (--threads 8)
	[ -f $ID.fastq ] || parallel-fastq-dump --sra-id $ID --threads 8
	
	# Perform QC with fastqc on the raw reads outputting everything in the fastqc folder (--outdir fastqc)
	fastqc --outdir 1_fastqc $ID.fastq
	
	# Remove fastq files to reduce disk usage
	# rm $ID.fastq
done

# do after aligning to add alignment information

mkdir 1_multiqc
multiqc --outdir 1_multiqc .

echo "loading script finished"
